### PR TITLE
WIP: Concurrent idris compilation

### PR DESCRIPTION
**Pull request by fabianhjr**
This is a work in progress to compile non-mutually depended upon modules concurrently.

Doesn't completely work right now, need to investigate more on Context and provide a way to merge two separate Contexts.

### CODE REVIEW

**Comment by fabianhjr** [comment]
There are two diverging approaches I am considering; either get the contents of the second context and add it to the first through a monadic traversal or create a pure Context -> Context -> Context merge.

By the looks of the record there is a Ref of an IOArray I would need to merge so I feel more inclined to do the first; Another issue is that it appears I will have to block while adding to the global context so that might slow the concurrency advantage a bit.

**Comment by fabianhjr** [comment]
This version compiles and works(tm) sometimes. (Flaky depending on Ctxt updates)

However seems like there is almost no performance gains. Will also investigate splitting the context to a local and global one.

**Comment by gallais** [comment]
> provide a way to merge two separate Contexts.

I expect it would make more sense to keep a single IOArray and
have a main thread owning it. The workers can then request fresh
pointers in an atomic way if they need to declare a new toplevel
definition.

Metas assignment should also be atomic and either succeed or return
the value already stored (so that you can compare it with the one you
wanted to put in and make sure they do match up).

File: libs/prelude/Prelude/Interfaces.idr [hunk 1]
```diff
@@ -174,12 +174,24 @@ interface Applicative m => Monad m where
   (>>=) x f = join (f <$> x)
   join x = x >>= id
 
-%allow_overloads (>>=)
+public export
+(=<<) : Monad m => (a -> m b) -> m a -> m b
```
Comment (Russoul) [inline libs/prelude/Prelude/Interfaces.idr:178]: This one is already defined at: contrib/Control/Monad/Syntax.idr

File: libs/prelude/Prelude/Ops.idr [hunk 2]
```diff
@@ -15,6 +15,7 @@ infixr 7 ::, ++
 
 -- Functor/Applicative/Monad/Algebra operators
 infixl 1 >>=, >>
+infixr 1 =<<, <<
```
Comment (Russoul) [inline libs/prelude/Prelude/Ops.idr:18]: (=<<)

File: src/Core/Core.idr [hunk 3]
```diff
@@ -418,70 +418,37 @@ export
 %inline
 coreLift : IO a -> Core a
 coreLift op = MkCore (do op' <- op
-                         pure (Right op'))
-
-{- Monad, Applicative, Traversable are specialised by hand for Core.
```
Comment (Russoul) [inline src/Core/Core.idr:423]: Are we there yet ? I know that the partial evaluator is a bit broken. But, on the other hand, there is machinery that transforms parametric method calls to specialised versions. The question is: does it work well ? Better ask @edwinb :)

File: libs/prelude/Prelude/Interfaces.idr (hunk 1 above)
Comment (fabianhjr) [reply libs/prelude/Prelude/Interfaces.idr:178]: Thanks for the heads up, added `<<` to the same Control.Monad.Syntax file and removed both flipped additions on prelude but moved this to #856 to separate the despecialization from the concurrency bit.

File: src/Core/Core.idr (hunk 3 above)
Comment (fabianhjr) [reply src/Core/Core.idr:423]: On testing it went from make of 114 seconds to 115 seconds. Separating this commit into #856)

**Comment by andrevidela** [comment]
Ping, @fabianhjr are you still working on this?

**Comment by fabianhjr** [comment]
Yes, but I went down a huge rabbit hole splitting Context unu

**Pull request closed without merging**